modules/sr/robot3/arduino_devices.py

```diff
@@ -0,0 +1,234 @@
+        """
+        self._check_if_disabled()
+        if self.mode not in DIGITAL_WRITE_MODES:
+            raise IOError(f'Digital write is not supported in {self.mode}')
+
+        self._digital_write(value)
+
+    def _analogue_read(self) -> float:
+        return map_to_range(0, 1, *self._ANALOGUE_RANGE, random.random())
+
+    def analog_read(self) -> float:
+        """
+        Get the analogue voltage on the pin.
+
+        This is returned in volts. Only pins A0-A7 support analogue reads.
+
+        :raises IOError: If the pin or its current mode does not support analogue read.
+        :raises IOError: If this pin cannot be controlled.
+        :return: The analogue voltage on the pin, ranges from 0 to 5.
+        """
+        self._check_if_disabled()
+        if self.mode not in ANALOG_READ_MODES:
+            raise IOError(f'Analogue read is not supported in {self.mode}')
+        if not self._supports_analogue:
+            raise IOError('Pin does not support analogue read')
+
+        return add_jitter(self._analogue_read(), *self._ANALOGUE_RANGE)
+
+
+class DisabledPin(PinDevice):
+    def __init__(self) -> None:
+        super().__init__(supports_analogue=False, disabled=True)
+
+
+class EmptyPin(PinDevice):
+    pass
+
+
+class DistanceSensor(PinDevice):
+    """
+    A standard Webots distance sensor, adapted to being a voltage based arduino sensor.
+    """
+
+    def __init__(self, webot: Robot, sensor_name: str) -> None:
+        super().__init__(supports_analogue=True)
+        self.webot_sensor = get_robot_device(webot, sensor_name, WebotsDistanceSensor)
+        self.webot_sensor.enable(int(webot.getBasicTimeStep()))
+
+    def _analogue_read(self) -> float:
+        return map_to_range(
+            self.webot_sensor.getMinValue(),
+            self.webot_sensor.getMaxValue(),
+            *self._ANALOGUE_RANGE,
+            self.webot_sensor.getValue(),
+        )
+
+
+class PressureSensor(PinDevice):
+    """
+    A Webots touch sensor with pressure, adapted to being a voltage based arduino sensor.
+    """
+
+    def __init__(self, webot: Robot, sensor_name: str) -> None:
+        super().__init__(supports_analogue=True)
+        self.webot_sensor = get_robot_device(webot, sensor_name, TouchSensor)
+        self.webot_sensor.enable(int(webot.getBasicTimeStep()))
+
+    def _analogue_read(self) -> float:
+        # Currently we only the return Z-axis force.
+        return self.webot_sensor.getValues()[2] / 100
+
+
+class Microswitch(PinDevice):
+    """
+    A standard Webots touch sensor.
+    """
+
+    def __init__(self, webot: Robot, sensor_name: str) -> None:
+        super().__init__(supports_analogue=False)
+        self.webot_sensor = get_robot_device(webot, sensor_name, TouchSensor)
+        self.webot_sensor.enable(int(webot.getBasicTimeStep()))
+
+    def _analogue_read(self) -> float:
+        return self._ANALOGUE_RANGE[int(self._digital_read())]
```

Comment:
Why is this defined? The microswitch doesn't have an analog value

Reply:
d137ee83b920a3c8c0c0fa1a508b6d7548a0cb21